config/registry.yaml

```diff
@@ -12,3 +12,13 @@ registry:
     collections: lsst.daf.butler.registry.collections.synthIntKey.SynthIntKeyCollectionManager
     datasets: lsst.daf.butler.registry.datasets.byDimensions.ByDimensionsDatasetRecordStorageManager
     datastores: lsst.daf.butler.registry.bridge.monolithic.MonolithicDatastoreRegistryBridgeManager
+  schema_versions:
+    core:
+      version: 0.0.1
+      digest: ""
+    dimensions:
+      version: 0.0.1
```

Comment:
Does this have to match the version number in dimensions.yaml?

Reply:
This is the question that I forgot to ask with all other complications. I am not sure what how these versions relate to each other but we should try to un-confuse things here. I know that dimensions.yaml version is used internally and we can probably have multiple versions (and I think it's also a single integer number). 

 @TallJimbo, what are your thoughts on how that version is related to schema version? Do we actually need that one?

Reply:
I think it'd probably be ideal if the version in `dimensions.yaml` could be the only one, and also serve as a schema version.  We could also probably find a way to pass a version defined in this part of the config into the dimensions system, but I'm guessing it would require adding it as an argument to a bunch of constructors.  I definitely think we should have only one version number for dimensions.

I would be quite happy to use the `VersionTuple` class introduced on this ticket for the version information stored in `DimensionUniverse`, if we can organize the package structure to make that dependency okay.

As a side thought, I know the dimensions system needs some work to make backwards compatibility easier; very small changes (like adding a new HTM level) right now will make a totally new universe that knows nothing about the old one, and since dimensions are used to define tables (e.g. `dataset_collection_XXXX`) for some managers, that's a real problem.  The (vague) ideas I have for fixing this include:
 - Start including the dimension version number into `DimensionGraph.encode` output so we can at least tell when a `dataset_collection_XXXX` table was defined with a different dimension version than we have.
 - Have `dimensions.yaml` include multiple versions (under different `version: ` headers) of the complete dimensions definition.  Until a repo is fully migrated away from an old version, any older dimensions definitions used in that repo must be present in configuration as well.  If necessary we could add some kind of inheritance/override stuff to configs to reduce the duplication this implies.
 - Provide a way to translate a `DimensionGraph` from one version to another, by looking up the `DimensionUniverse` for both versions.  This would need to assume that the translated dimensions are present in both and have the same relationships, but could otherwise differ.

That's probably mostly not relevant for this ticket (it's more about making more changes into patch or minor changes instead of forcing all changes to be major changes), but this discussion reminded me that I wanted to discuss these ideas with you, and it gives you an idea of what the version in `dimensions.yaml` might do.

Reply:
I'm OK with using version from `dimensions.yaml` but I worry that we may be overloading that version number with different purposes. The schema version number should reflect the database side of things, and that is an implementation detail and it could change if you use different implementation of a manager class for example. This could mean that we have to switch to incompatible version number even if `dimensions.yaml` does not change. I'm not entirely sure but I think that existing version number has some different purpose which does not depend on implementation detail like database schema.

Reply:
It's a good point that the version in `dimensions.yaml` is not quite a schema version.  But I also think it's all we need (pending the question of versioning managers that's being discussed on Jira) - I think we can say that the combination of the `dimensions.yaml` version with versions that would be embedded in the daf_butler codebase (I'm imagining an overall daf_butler schema version + a dimension manager version, but that's not the only possibility) totally controls the dimensions part of the schema.

Or more precisely, there isn't a well-defined "dimensions part of the schema", but if we think about the full schema "version" being an unorderable hash of many different human-meaningful identifiers (see my last Jira comment), then the `dimensions.yaml` version number is still a sufficient number to capture any changes in that file/yaml-section, even if it has to be hashed together with other in-code versions and config labels to map to the full schema.

Reply:
One more thought on dimensions - encoding of the dimension graph into `dataset_collection_XXXX` table names makes it very fragile w.r.t. adding or removing of the dimensions to/from the universe. There may be (complicated) workaround for that by adding a separate registry/table for mapping a set of dimensions to table name (or just a serial table number). It will be append-only and can be cached efficiently but it does add some extra logic.

Reply:
I was already planning to follow up (on DM-24660, or maybe DM-24613 to do it right the first time for quantum tables) on your suggestion on a previous ticket review to start saving the full table names into the `dataset_type` table, instead of relying on the encoding producing the same thing each time.  I think if we encode the dimension version number, too, that may be enough to avoid that kind of logic, but we'll see.